store/tooling/auth.go

```diff
@@ -0,0 +1,257 @@
+		return &SimpleCreds{
+			Scheme: creds.Scheme,
+			Value:  creds.Value,
+		}, nil
+	}
+
+	return &UbuntuOneCreds{User: auth.UserState{
+		StoreMacaroon:   creds.Macaroon,
+		StoreDischarges: creds.Discharges,
+	}}, nil
+}
+
+func parseAuthData(data []byte, what string, parsers ...parseAuthFunc) (*authData, error) {
+	var firstErr error
+	for _, p := range parsers {
+		parsed, likely, err := p(data, what)
+		if err == nil {
+			return parsed, nil
+		}
+		if likely && firstErr == nil {
+			firstErr = err
+		}
+	}
+	if firstErr == nil {
+		return nil, fmt.Errorf("invalid auth %s: not a recognizable format", what)
+	}
+	return nil, firstErr
+}
+
+func parseAuthJSON(data []byte, what string) (*authData, bool, error) {
+	var creds struct {
+		Macaroon   string   `json:"macaroon"`
+		Discharges []string `json:"discharges"`
+	}
+	err := json.Unmarshal(data, &creds)
+	if err != nil {
+		likely := data[0] == '{'
+		return nil, likely, fmt.Errorf("cannot decode auth %s: %v", what, err)
+	}
+	if creds.Macaroon == "" || len(creds.Discharges) == 0 {
+		return nil, true, fmt.Errorf("invalid auth %s: missing fields", what)
+	}
+	return &authData{
+		Macaroon:   creds.Macaroon,
+		Discharges: creds.Discharges,
+	}, false, nil
+}
+
+func snapcraftLoginSection() string {
+	if snapdenv.UseStagingStore() {
+		return "login.staging.ubuntu.com"
+	}
+	return "login.ubuntu.com"
+}
+
+// parseSnapcraftLoginFile parses the content of snapcraft <v7 exported
+// login credentials files.
+func parseSnapcraftLoginFile(data []byte, what string) (*authData, bool, error) {
+	errPrefix := fmt.Sprintf("invalid snapcraft login %s", what)
+
+	cfg := goconfigparser.New()
+	// XXX this seems to almost always succeed
+	if err := cfg.ReadString(string(data)); err != nil {
+		likely := data[0] == '['
+		return nil, likely, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	sec := snapcraftLoginSection()
+	macaroon, err := cfg.Get(sec, "macaroon")
+	if err != nil {
+		return nil, true, fmt.Errorf("%s: %s", errPrefix, err)
+	}
+	unboundDischarge, err := cfg.Get(sec, "unbound_discharge")
+	if err != nil {
+		return nil, true, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	if macaroon == "" || unboundDischarge == "" {
+		return nil, true, fmt.Errorf("%s: empty fields", errPrefix)
+	}
+	return &authData{
+		Macaroon:   macaroon,
+		Discharges: []string{unboundDischarge},
+	}, false, nil
+}
+
+// parseAuthBase64JSON parses snapcraft v7+ base64-encoded auth credential data.
+func parseAuthBase64JSON(data []byte, what string) (*authData, bool, error) {
+	jsonData := make([]byte, base64.StdEncoding.DecodedLen(len(data)))
+	dataLen, err := base64.StdEncoding.Decode(jsonData, data)
+	if err != nil {
+		return nil, false, fmt.Errorf("cannot decode base64-encoded auth %s: %v", what, err)
+	}
+	var m map[string]interface{}
+	if err := json.Unmarshal(jsonData[:dataLen], &m); err != nil {
+		return nil, true, fmt.Errorf("cannot unmarshal base64-decoded auth %s: %v", what, err)
+	}
+	r, _ := m["r"].(string)
```

Comment:
Silly question, what does "r","d","t" stand for here?

Reply:
t is for type, r is for root, and d is for discharge afaiu, it's using single letter field names for compactness like some of the JOSE/JWT specs